modules/deployment-model-impl/src/main/java/org/mule/runtime/module/deployment/impl/internal/maven/AbstractMavenClassLoaderModelLoader.java

```diff
@@ -233,19 +238,53 @@ private ClassLoaderModel createLightPackageClassLoaderModel(File artifactFile, M
                                        new HashSet<>(getAttribute(attributes, PRIVILEGED_ARTIFACTS_IDS)))
           .exportingResources(new HashSet<>(getAttribute(attributes, EXPORTED_RESOURCES)))
           .includeTestDependencies(valueOf(getSimpleAttribute(attributes, INCLUDE_TEST_DEPENDENCIES, "false")));
-      Set<BundleDependency> bundleDependencies =
-          dependencies.stream().filter(mavenClientDependency -> !mavenClientDependency.getScope().equals(PROVIDED))
-              .map(mavenClientDependency -> convertBundleDependency(mavenClientDependency)).collect(toSet());
-      loadUrls(artifactFile, classLoaderModelBuilder, bundleDependencies);
-      Set<BundleDependency> allBundleDependencies =
-          dependencies.stream().map(mavenClientDependency -> convertBundleDependency(mavenClientDependency)).collect(toSet());
-      classLoaderModelBuilder.dependingOn(allBundleDependencies);
+      Set<BundleDependency> missingApiDependencyBundles = findMissingApiDependencies(dependencies, attributes);
+      Stream<BundleDependency> bundleDependencies = dependencies.stream()
+          .filter(mavenClientDependency -> !mavenClientDependency.getScope().equals(PROVIDED))
+          .map(this::convertBundleDependency);
+      loadUrls(artifactFile, classLoaderModelBuilder, concat(bundleDependencies,
+                                                             missingApiDependencyBundles.stream()).collect(toSet()));
+      Stream<BundleDependency> allBundleDependencies = dependencies.stream().map(this::convertBundleDependency);
+      classLoaderModelBuilder.dependingOn(concat(allBundleDependencies,
+                                                 missingApiDependencyBundles.stream()).collect(toSet()));
       return classLoaderModelBuilder.build();
     } finally {
       deleteQuietly(temporaryDirectory);
     }
   }
 
+  private Set<BundleDependency> findMissingApiDependencies(List<org.mule.maven.client.api.model.BundleDependency> dependencies,
+                                                           Map<String, Object> attributes) {
+    Set<org.mule.maven.client.api.model.BundleDependency> apiDependencies = dependencies.stream()
+        .filter(dependency -> {
+          Optional<String> classifier = dependency.getDescriptor().getClassifier();
+          return classifier.isPresent() && API_CLASSIFIERS.contains(classifier.get());
+        })
+        .collect(toSet());
+
+    Deque<org.mule.maven.client.api.model.BundleDependency> dependenciesToCheck = new ArrayDeque<>(apiDependencies);
+    Set<org.mule.maven.client.api.model.BundleDependency> missingApiDependencies = new HashSet<>();
+
+    while (!dependenciesToCheck.isEmpty()) {
```

Comment:
Just realize we have almost the algorithm at org.mule.runtime.deployment.model.internal.plugin.BundlePluginDependenciesResolver#resolvePluginsDependencies
We should refactor all of this and have it in one resolver.